drivers/stepper/adi_tmc/bus/adi_tmc_uart.c

```diff
@@ -0,0 +1,171 @@
+/* TMC UART protocol constants */
+#define TMC_UART_SYNC_BYTE              0x05
+#define TMC_UART_WRITE_BIT              0x80
+
+uint8_t tmc_uart_calc_crc(const uint8_t *datagram, uint8_t len)
+{
+	uint8_t crc = 0;
+
+	for (uint8_t i = 0; i < len; i++) {
+		uint8_t current_byte = datagram[i];
+
+		for (uint8_t j = 0; j < 8; j++) {
+			if ((crc >> 7) ^ (current_byte & 0x01)) {
+				crc = (crc << 1) ^ 0x07;
+			} else {
+				crc = (crc << 1) & 0xFF;
+			}
+			current_byte = current_byte >> 1;
+		}
+	}
+
+	return crc;
+}
+
+int tmc_uart_write_register(const struct device *uart, uint8_t slave_addr, uint8_t register_address,
+			    uint32_t data)
+{
+	uint8_t buffer[TMC_UART_DATAGRAM_SIZE];
+	int err = 0;
+	uint8_t echo_byte;
+
+	/* Format the UART TMC datagram */
+	buffer[0] = TMC_UART_SYNC_BYTE;                    /* Sync byte */
+	buffer[1] = slave_addr;                            /* Slave address */
+	buffer[2] = register_address | TMC_UART_WRITE_BIT; /* Register address with write bit set */
+	buffer[3] = (data >> 24) & 0xFF;                   /* Data byte 3 (MSB) */
+	buffer[4] = (data >> 16) & 0xFF;                   /* Data byte 2 */
+	buffer[5] = (data >> 8) & 0xFF;                    /* Data byte 1 */
+	buffer[6] = data & 0xFF;                           /* Data byte 0 (LSB) */
+	buffer[7] = tmc_uart_calc_crc(buffer, sizeof(buffer) - 1U); /* CRC */
+
+	/* Send datagram byte by byte using polling */
+	for (size_t i = 0; i < sizeof(buffer); i++) {
+		uart_poll_out(uart, buffer[i]);
+
+		/* Wait for echo with timeout */
+		k_timepoint_t end = sys_timepoint_calc(K_MSEC(5)); /* 5ms timeout for echo */
+
+		do {
+			err = uart_poll_in(uart, &echo_byte);
+			if (err >= 0 && echo_byte == buffer[i]) {
+				/* Received matching echo */
+				break;
+			}
+		} while (err == -1 && !sys_timepoint_expired(end));
+
+		if (err < 0 || echo_byte != buffer[i]) {
+			LOG_WRN("Echo mismatch or timeout for byte %d: sent 0x%02X", i, buffer[i]);
+			return -EIO;
+		}
+	}
+
+	return 0;
+}
+
+int tmc_uart_read_register(const struct device *uart, uint8_t slave_addr, uint8_t register_address,
+			   uint32_t *data)
+{
+	uint8_t write_buffer[TMC_UART_READ_REQ_DATAGRAM_SIZE]; /* Only need 4 bytes for a read */
+							       /* request */
+	uint8_t read_buffer[TMC_UART_DATAGRAM_SIZE];
+	struct uart_config uart_cfg;
+
+	int err = 0;
+
+	/* Get current UART configuration */
+	err = uart_config_get(uart, &uart_cfg);
+	if (err) {
+		return err;
+	}
+
+	/* Calculate delay based on UART baudrate (SENDDELAY: default=8 bit times))*/
+	uint32_t delay_us = (8 * 1000000) / uart_cfg.baudrate;
+
+	/* Format the UART TMC datagram for read */
+	write_buffer[0] = TMC_UART_SYNC_BYTE; /* Sync byte */
+	write_buffer[1] = slave_addr;         /* Slave address */
+	write_buffer[2] = register_address;   /* Register address without write bit */
+	write_buffer[3] = tmc_uart_calc_crc(write_buffer, 3);
+	uint8_t echo_byte;
+
+	/* Send read request byte by byte and wait for each echo */
```

Comment:
Extract to a function, same code as in the write method

Reply:
Good suggestion. Moved it to a new function, `tmc_uart_send_byte_with_echo`. Let me know if this is fine.